### PR TITLE
add capability for sending bug reports

### LINKED ISSUE #49542: Send diagnostics tool info to Slack
Whenever clicking the "Report bug" button (cf. #49543) we should send to Slack a summary message along with the JSON file currently generated by the diagnostics tool.

### PR DESCRIPTION
**Pull request by filiphric**
Closes https://github.com/metabase/metabase/issues/49542

### Description

Add a button to send diagnostic data to slack. Right now, there’s an option to download the data as a json when cmd/ctrl + F1 key is pressed. View [this project](https://github.com/orgs/metabase/projects/102/views/1) for more details. This feature is only available on stats.metabase.com or localhost.

### How to verify

1. cmd/ctrl + F1
2. "send to slack" button appears
3. hit the button and see a report appear in #bug-reports-internal in Slack metabase

### Tests
- bug report channel input appears on Slack integration when MB_BUG_REPORTER_ENABLED is set
- bug report channel input field is not visible when MB_BUG_REPORTER_ENABLED is not set
- “send to slack” button appears when all of these are true:
    - bug report channel is set in slack settings
    - MB_BUG_REPORTER_ENABLED env variable is set up
- “send to slack” button does not appear when either one of the conditions is missing
- “send to slack” button shows loading 
- “send to slack” button hides loading when action is done
- notification message is shown when sending to slack is successful and unsuccessful
- diagnostic data can be downloaded
- diagnostic data is sent to files channel
- diagnostic data summary with original file attached is sent to bug reports channel

### CODE REVIEW

File: rspack.config.js
Comment (iethree) [inline rspack.config.js:289]: instead of a build-time frontend environment variable, we should use a runtime backend environment variable

File: frontend/src/metabase/env.ts
Comment (iethree) [inline frontend/src/metabase/env.ts:18]: This would require us to a build a separate bundle to enable this - I think instead we should use an env var on the server to turn it on

File: src/metabase/integrations/slack.clj
Comment (iethree) [inline src/metabase/integrations/slack.clj:91]: we should add a backend setting for `slack-bug-reports-enabled` and have it settable only by environment variable, and use that setting to toggle visibility of the bug report channel input on the frontend

File: frontend/src/metabase/components/ErrorPages/ErrorDiagnosticModal.tsx
Comment (iethree) [inline frontend/src/metabase/components/ErrorPages/ErrorDiagnosticModal.tsx:100]: we should define this new endpoint as an RTK query api endpoint in `frontend/src/api`

File: frontend/src/metabase/components/ErrorPages/ErrorDiagnosticModal.tsx
Comment (iethree) [inline frontend/src/metabase/components/ErrorPages/ErrorDiagnosticModal.tsx:171]: Since this is a 1-shot bug report, I think from a ux standpoint, we need to add a `textarea` to describe the bug. I actually found a bug while testing this and reported it.. but nobody would ever have a clue what the bug was from the report


Screen Shot 2024-11-14 at 7 53 55 AM

File: src/metabase/integrations/slack.clj
Comment (iethree) [inline src/metabase/integrations/slack.clj:316]: i'm no clojurian, but maybe `set-bug-report-channel` would be more descriptive here?

**Review by iethree — CHANGES_REQUESTED** [review]
Very nice! This will be super useful.

A few requests:
1) instead of using a frontend build-time env var, we should use a backend setting that can be configured by env var
2) we should add an optional text field to give some context to the bug
3) we also need a backend review from a backend developer (like @piranha)

File: frontend/src/metabase/components/ErrorPages/ErrorDiagnosticModal.tsx
Comment (thebiglabasky) [reply frontend/src/metabase/components/ErrorPages/ErrorDiagnosticModal.tsx:167]: Yep that's planned!

File: frontend/src/metabase/api/bug-report.ts
Comment (iethree) [inline frontend/src/metabase/api/bug-report.ts:26]: you should not need another api endpoint to get a setting. you should be able to use

```ts
const isBugReportingEnabled = useSetting('bug-reporting-enabled');
```

File: src/metabase/api/util.clj
Comment (iethree) [inline src/metabase/api/util.clj:108]: I am very much out of my depth here, but I think you want to `defsetting` and make a real setting for bug-reporting enabled, so it can be accessed with other settings.

**Review by iethree — CHANGES_REQUESTED** [review]
Good changes, still some small things to align with how we handle settings

File: src/metabase/api/util.clj
Comment (piranha) [reply src/metabase/api/util.clj:108]: yeah, for sure, no sense imitating that with a private var

**Review by iethree — APPROVED** [review]
Frontend looks good to me 🥳 

Please wait for a backend review before merging

File: src/metabase/api/slack.clj
Comment (escherize) [inline src/metabase/api/slack.clj:18]: This comment would work better as a docstring, e.g.:

```suggestion
;; 
(defn- create-slack-message-blocks 
  "Create blocks for the Slack message with diagnostic information"
  [diagnostic-info file-info]
```

File: src/metabase/api/slack.clj
Comment (escherize) [inline src/metabase/api/slack.clj:76]: `:maybe` means nillable, so we can drop the `:or nil?` part:

```suggestion
   slack-bug-report-channel [:maybe :string]}
```

File: src/metabase/integrations/slack.clj
Comment (escherize) [inline src/metabase/integrations/slack.clj:319]:
```suggestion
  throws an error that advises an admin to create it."
```

File: src/metabase/public_settings.clj
Comment (escherize) [inline src/metabase/public_settings.clj:1068]: These 2 lines are unnecessary, and you can remove them because the environment lookup behavior will happen automatically by default. see: https://github.com/metabase/metabase/blob/master/src/metabase/models/setting.clj#L1162-L1172

File: src/metabase/api/slack.clj
Comment (escherize) [inline src/metabase/api/slack.clj:105]: `(when (not ...` -> `when-not`

```suggestion
        (when-not (slack/channel-exists? processed-files-channel)
```

File: src/metabase/integrations/slack.clj
Comment (escherize) [inline src/metabase/integrations/slack.clj:439]: I'll leave it up to you, but what do you think about updating function signature to accept a map in the 3rd arg position, with optional `:blocks` and `:attachments` keys? Overloading the 3rd arg position to be attachment data or blocks seems to complect the two separate concepts.

That would improve the readability when computing params in the `cond->` on 441.

File: src/metabase/integrations/slack.clj
Comment (escherize) [inline src/metabase/integrations/slack.clj:436]:
```suggestion
   Can accept `attachments` data or `blocks` as a vector, or as a map in the `:blocks` key."
```

**Review by escherize — CHANGES_REQUESTED** [review]
Thanks for this, I can't wait to get our error messages into slack!

I had a couple notes, mostly about some metabasey code style things.

I am curious to hear your take on changing how we handle `attachments-or-blocks` in `post-chat-message!`.

File: src/metabase/api/slack.clj
Comment (escherize) [inline src/metabase/api/slack.clj:17]: I'd remove this comment

File: src/metabase/public_settings.clj
Comment (escherize) [reply src/metabase/public_settings.clj:1068]: Looking at this again, I see that `:mb-bug-reporting-enabled` is different than `bug-report-enabled`. This will work by default if the name of the settings is `bug-reporting-enabled`. We usually don't allow a mismatch between the name of the defsetting and the env var used to control them.

Comment (escherize) [reply src/metabase/public_settings.clj:1068]: so if the env var to control this should be `MB_BUG_REPORTING_ENABLED`, I think this defestting should be called `bug-reporting-enabled`.

**Review by escherize — APPROVED** [review]
Thanks!

I'll approve this now, but I think we should align the defsetting name with the env var used to control it.

File: frontend/src/metabase/admin/settings/slack/components/SlackForm/SlackForm.tsx
Comment (ranquild) [inline frontend/src/metabase/admin/settings/slack/components/SlackForm/SlackForm.tsx:24]: @filiphric fyi you can set `nullable` property on `FormInput` to do that

File: frontend/src/metabase/components/ErrorPages/ErrorDiagnosticModal.tsx
Comment (ranquild) [inline frontend/src/metabase/components/ErrorPages/ErrorDiagnosticModal.tsx:43]: `useSendBugReportMutation` returns the fetch status, so this is not needed

**Pull request merged**
